provider/packet/environ.go

```diff
@@ -0,0 +1,571 @@
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	d, err = waitDeviceActive(e.packetClient, d.ID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	inst := &packetDevice{e, d}
+	amd64 := arch.AMD64
+	mem, err := strconv.ParseUint(d.Plan.Specs.Memory.Total[:len(d.Plan.Specs.Memory.Total)-2], 10, 64)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var cpus uint64
+	if len(inst.Plan.Specs.Cpus) > 0 {
+		cpus = uint64(inst.Plan.Specs.Cpus[0].Count)
+	}
+
+	hc := &instance.HardwareCharacteristics{
+		Arch: &amd64,
+		Mem:  &mem,
+		// RootDisk: &instanceSpec.InstanceType.RootDisk,
+		CpuCores: &cpus,
+	}
+
+	return &environs.StartInstanceResult{
+		Instance: inst,
+		Hardware: hc,
+	}, nil
+}
+
+// supportedInstanceTypes returns the instance types supported by packet.
+func (e *environ) supportedInstanceTypes() ([]instances.InstanceType, error) {
+	options := &packngo.ListOptions{
+		Includes: []string{"available_in"},
+	}
+	plans, _, err := e.packetClient.Plans.List(options)
+
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving supported instance types")
+	}
+
+	var instTypes []instances.InstanceType
+nextPlan:
+	for _, plan := range plans {
+		if !validPlan(plan) {
+			continue
+		}
+
+		if !isPlanSupported(plan, e.cloud.Region) {
+			continue
+		}
+
+		var instArch string
+		switch {
+		case strings.HasSuffix(plan.Name, ".x86"):
+			instArch = arch.AMD64
+		case strings.HasSuffix(plan.Name, ".arm"):
+			instArch = arch.ARM64
+		default:
+			continue nextPlan
+
+		}
+
+		mem, err := parseMemValue(plan.Specs.Memory.Total)
+		if err != nil {
+			continue
+		}
+
+		instTypes = append(instTypes,
+			instances.InstanceType{
+				Id:       plan.ID,
+				Name:     plan.Name,
+				CpuCores: uint64(plan.Specs.Cpus[0].Count),
+				Mem:      mem,
+				Arches:   []string{instArch},
+				// Scale per hour costs so they can be represented as an integer for sorting purposes.
+				Cost: uint64(plan.Pricing.Hour * 1000.0),
+				// TODO: returned by packet's API but not exposed by the packngo client
+				// Deprecated: plan.Legacy,
+			})
+	}
+
+	return instTypes, nil
+}
+
+func isPlanSupported(plan packngo.Plan, code string) bool {
+	for _, f := range plan.AvailableIn {
+		if f.Code == code {
+			return true
+		}
+	}
+	return false
+}
+
+func validPlan(plan packngo.Plan) bool {
+	isInvalid := plan.Pricing == nil ||
+		plan.Specs == nil ||
+		plan.Specs.Memory == nil ||
+		len(plan.Specs.Cpus) == 0 || plan.Specs.Cpus[0].Count == 0
+
+	return !isInvalid
+}
+
+func parseMemValue(v string) (uint64, error) {
+	var scaler = uint64(1)
+	if strings.HasSuffix(v, "GB") {
+		scaler = 1024
+		v = strings.TrimSuffix(v, "GB")
+	}
+
+	val, err := strconv.ParseUint(v, 10, 64)
+	return val * scaler, err
+}
+
+func (e *environ) findInstanceSpec(controller bool, allImages []*imagemetadata.ImageMetadata, instanceTypes []instances.InstanceType, ic *instances.InstanceConstraint) (*instances.InstanceSpec, error) {
```

Comment:
@achilleasa I've extended this so it can filter out available plan in the given facility. 

I tried testing without provided `--bootstrap-series bionic` but it kept failing as it would upload focal agent to bionic instance. I was not able to determine where this goes wrong. Maybe you have an idea what is going on.

Reply:
If you don't provide ``--bootsrtap-series`, juju will opt for the default series (`focal` on develop). So, you probably need to check the code that populates the image list and see what image you eventually request from packet's API when spinning up instances.

Reply:
Is this in `bootstrapModelConfig` under `default-series:focal`?

Reply:
Under instance `ic *instances.InstanceConstraint`?

Reply:
Yes, it will get injected in there. So you need to check the code you added which builds the image list. Previously, you hardcoded the image that got passed to packet's API to bionic. Has something changed in that logic?

Reply:
I just want to make it dynamic based on the series that is being passed from bootstrap command.

Reply:
If you query packet's API and provide the full list of available images with their series details, the filtering code will automatically select the correct one for you based on all provided constraints. 

Building the image name dynamically is probably the wrong approach for this and will probably make it harder to debug the provider if/when the packet images get renamed at some point in the future.

Reply:
So i'm trying to avoid this
```
suitableImages := []*imagemetadata.ImageMetadata{
		{
			Id:   "ubuntu_18_04",
			Arch: arch.AMD64,
		},
	}
```
We had this before I want populate the list based on the series rather than having hardcoded values.

Does this make sense?

Reply:
To elaborate this a little bit more. So Equinix OSs have following properties:
```
{
    "name": "Ubuntu 20.10",
    "slug": "ubuntu_20_10",
    "distro": "ubuntu",
    "version": "20.10",
    "provisionable_on": [
      "c1.small.x86",
      "baremetal_1",
      "c2.medium.x86",
      "c3.medium.x86",
```
provisionable_on can be used to determine if the requested OS is available on selected plan which we get from the instanceTypes []instances.InstanceType `it.Name`. So based on that we could determine if the OS is available on that plan and pick appropriate ubuntu series (version). Using all this information we can push items to `suitableImages`.

Reply:
In that case, you can use https://github.com/juju/os/blob/master/series/supportedseries.go#L476 to map `version` into a suitable series name when processing the above list of OSes.